java/java.source.base/src/org/netbeans/modules/java/source/save/Reformatter.java

```diff
@@ -1151,7 +1151,7 @@ public Boolean visitVariable(VariableTree node, Void p) {
                                 newline();
                             else
                                 space();
-                        } else {
+                        } else if (sp.getStartPosition(root, mods) != sp.getStartPosition(root, node.getType())) {
```

Comment:
the implicit `final` would have the same start position as the type (which is the next item in the tree I guess) because its hidden?

I am wondering if this is just an impl detail or actually stable behavior, because there would be also the NOPOS constant
https://github.com/openjdk/jdk/blob/05d38604a2c620dcaf8682f02dae2fddab8e0c0b/src/jdk.compiler/share/classes/com/sun/source/util/SourcePositions.java#L41-L43
which could be potentially returned in future?

if implicit items have the same start and end pos we could extract this into:
```java
private static boolean isHidden(CompilationUnitTree cut, Tree tree) {
    long start = sp.getStartPosition(cut, tree);
    long end = sp.getEndPosition(cut, tree);
    return start == end && start != javax.tools.Diagnostic.NOPOS;
}
```
Maybe that would be less likely to break on a javac update (assuming it works)?

Reply:
> if implicit items have the same start and end pos

Thanks @mbien but this hits on the crux of the issue. The implicit `final` does not have the same start and end position (zero length), or report `NOPOS`. It reports the same start _and end_ positions as the following type element.  It has the same length and overlaps.

Were the element reported empty or at no position, it shouldn't get past the enclosing check at https://github.com/apache/netbeans/blob/master/java/java.source.base/src/org/netbeans/modules/java/source/save/Reformatter.java#L1146  I assume that the reported end position changed which started causing this issue, but haven't checked old versions to be sure.

Reply:
Bizarrely I also think that #4667 might be caused by this same issue.

Reply:
Ironically @jlahoda might have fixed this today? https://bugs.openjdk.org/browse/JDK-8293897

The hotfix shouldn't break if this is fixed. And it's covered by tests now I think. Could add a comment to revert when upstream fixed?

Reply:
>The hotfix shouldn't break if this is fixed. And it's covered by tests now I think. Could add a comment to revert when upstream fixed?

Yeah I think so - lets get this in. The tests are indeed important so that we notice regressions quickly. Good that you found that bug report, it clears things up.

If the upstream fix makes it into javac 19.0.1 too, we could adjust/partially revert it later while upgrading nb-javac.